controllers/provider-alicloud/pkg/controller/worker/machines.go

```diff
@@ -40,13 +44,94 @@ func (w *workerDelegate) MachineClassList() runtime.Object {
 	return &machinev1alpha1.AlicloudMachineClassList{}
 }
 
+// getSeedCloudProviderCredentials gets Seed's cloud provider's credentials from BackupBucket SecretRef
+func (w *workerDelegate) getSeedCloudProviderCredentials(ctx context.Context) (*alicloud.Credentials, error) {
+	backupBuckets := &extensionsv1alpha1.BackupBucketList{}
+	err := w.client.List(ctx, backupBuckets)
+	if err != nil {
+		return nil, err
+	}
+	credentials, err := alicloud.ReadCredentialsFromSecretRef(ctx, w.client, &backupBuckets.Items[0].Spec.SecretRef)
+	if err != nil {
+		return nil, err
+	}
+	return credentials, nil
+}
+
+// ShareMachineImages checks whether Shoot's Alicloud account has permissions to use the machine images. If it can't
+// access them, these images will be shared with it from Seed's Alicloud account.
+func (w *workerDelegate) ShareMachineImages(ctx context.Context) error {
+	var (
+		shootAlicloudSTSClient      alicloudclient.STS
+		shootCloudProviderAccountID string
+		err                         error
+		logger                      logr.Logger
+	)
+	shootAlicloudSTSClient = nil
+	shootCloudProviderAccountID = ""
+	logger = log.Log.WithName("worker-actuator-image-share")
+
+	logger.Info("Reading Seed's cloud provider credentials", "worker", fmt.Sprintf("%s/%s", w.worker.Namespace, w.worker.Name))
+	seedCloudProviderCredentials, err := w.getSeedCloudProviderCredentials(ctx)
+	if err != nil {
+		return err
+	}
+	logger.Info("Reading Shoot's cloud provider credentials", "worker", fmt.Sprintf("%s/%s", w.worker.Namespace, w.worker.Name))
+	shootCloudProviderCredentials, err := alicloud.ReadCredentialsFromSecretRef(ctx, w.client, &w.worker.Spec.SecretRef)
+	if err != nil {
+		return err
+	}
+	logger.Info("Sharing the image with Shoot's alicloud account", "worker", fmt.Sprintf("%s/%s", w.worker.Namespace, w.worker.Name))
+	for _, machineclass := range w.machineClasses {
+		region := machineclass["region"].(string)
+		imageID := machineclass["imageID"].(string)
+
+		if shootAlicloudSTSClient == nil || shootCloudProviderAccountID == "" {
+			shootAlicloudSTSClient, err = alicloudclient.NewSTSClient(ctx, region, shootCloudProviderCredentials.AccessKeyID, shootCloudProviderCredentials.AccessKeySecret)
+			if err != nil {
+				return err
+			}
+			shootCloudProviderAccountID, err = shootAlicloudSTSClient.GetAccountIDFromCallerIdentity(ctx)
+			if err != nil {
+				return err
+			}
+		}
+
+		seedAlicloudECSClient, err := alicloudclient.NewECSClient(ctx, region, seedCloudProviderCredentials.AccessKeyID, seedCloudProviderCredentials.AccessKeySecret)
```

Comment:
same, move out of the `for` loop. Also, when following above suggestion, you can even instantiate the alicloud client for the seed account only once instead of in every reconciliation loop

Reply:
done